google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/DefaultCloudSdkService.java

```diff
@@ -40,4 +83,206 @@ public String getCloudSdkHomePath() {
   public void setCloudSdkHomePath(String cloudSdkHomePath) {
     propertiesComponent.setValue(CLOUD_SDK_PROPERTY_KEY, cloudSdkHomePath);
   }
+
+  @Nullable
+  private Path getJavaToolsBasePath() {
+    return getCloudSdkHomePath() != null
+        ? Paths.get(getCloudSdkHomePath(), JAVA_TOOLS_RELATIVE_PATH.toString())
+        : null;
+  }
+
+  @Nullable
+  @Override
+  public File getToolsApiJarFile() {
+    return getJavaToolsBasePath() != null
+        ? getJavaToolsBasePath().resolve(
+            JpsAppEngineModuleExtensionImpl.LIB_APPENGINE_TOOLS_API_JAR).toFile()
+        : null;
+  }
+
+  @NotNull
+  @Override
+  public File[] getLibraries() {
+    return getJavaToolsBasePath() != null
+        ? getJarsFromDirectory(getJavaToolsBasePath().resolve(Paths.get("lib", "shared")).toFile())
+        : new File[0];
+  }
+
+  @NotNull
+  @Override
+  public File[] getJspLibraries() {
+    return getJavaToolsBasePath() != null
+        ? getJarsFromDirectory(
+            getJavaToolsBasePath().resolve(Paths.get("lib", "shared", "jsp")).toFile())
+        : new File[0];
+  }
+
+  @Override
+  public boolean isMethodInBlacklist(@NotNull String className, @NotNull String methodName) {
+    if (myMethodsBlackList == null) {
+      try {
+        myMethodsBlackList = loadBlackList();
+      } catch (IOException ioe) {
+        logger.error(ioe);
+        myMethodsBlackList = new THashMap<>();
+      }
+    }
+    final Set<String> methods = myMethodsBlackList.get(className);
+    return methods != null && methods.contains(methodName);
+  }
+
+  @Override
+  public boolean isClassInWhiteList(@NotNull String className) {
+    return AppEngineJreWhitelist.contains(className);
+  }
+
+  @NotNull
+  @Override
+  public List<String> getUserLibraryPaths() {
+    List<String> result = new ArrayList<>();
+    String libUserDirectory = getLibUserDirectoryPath();
+    if (libUserDirectory != null && new File(libUserDirectory).exists()) {
+      result.add(libUserDirectory);
+    }
+
+    if (getJavaToolsBasePath() == null) {
```

Comment:
If javaToolsBasePath is null it will return an empty list which was just to satisfy the caller that will NPE otherwise. But I suppose this should be refactored.